scripts/firedrake-install

```diff
@@ -1462,9 +1462,10 @@ for opt in ["mpicc", "mpicxx", "mpif90", "mpiexec"]:
         os.remove(dest)
     if should_link:
         log.debug("Creating a wrapper for %s from %s to %s" % (opt, src, dest))
-        with open(dest, "w") as f:
+        # use low-level os.open to ensure permissions (incl. exec.) are set according to umask
```

Comment:
I might be being thick but I don't understand this comment - what's umask?

Reply:
The system that unix-like operating systems use to determine the default permissions of files on creation.

Reply:
umask is what determines the default permissions for newly created files and directories in a unix-like environment, to be more precise umask tells what bits should *not* be set. On a lot of machines this is set to 022 which means disallow writing for "group" and "other". I will admit I never really knew how this worked until a week ago.

To give more detail on the comment: if you do the standard `open('filename', 'w')` in python it does a lot of things for you: it first checks whether the file actually exists and if so, if you have permission to write to it, if not it will create a new file with some permissions (which importantly here does not include executable permission) both cases create a OS level file handle that can be writtent to. Then, as a second step it will also create a python file object that is setup to write to. If you want more control, you can split these two steps but you have to be more explicit. Here we already know we want to create a new file (because we have just deleted any existing version) and we want it to be writable. The default permissions are 0o777, i.e. all permissions, read write and execute for user, group and other, except those given by umask. That is precisely what you need for executable script: with a standard 022 umask you get a 755 == rwxr-xr-x file.

Alternatively you can open a file in the more standard way and fix permissions afterwards, with a chmod as was done previously, but it turns out to do that correctly is a bit awkward. To ask what the default permissions should be, python's os.umask requires you to actually change the value and it will return the current value. So you then have to set it back. And then you have to do the magic of something like `0o755 | ~umask` to filter out the disallowed permission from what you actually want.

Reply:
I understand, thank you!